apps/server/src/modules/tldraw/service/tldraw.ws.service.spec.ts

```diff
@@ -0,0 +1,376 @@
+				throw new Error('error');
+			});
+			const doc = new WsSharedDocDo('TEST', service);
+			const { msg } = createMessage([0, 1]);
+
+			return {
+				sendSpy,
+				doc,
+				msg,
+			};
+		};
+
+		it('should not call send method', async () => {
+			const { sendSpy, doc, msg } = await setup();
+
+			service.messageHandler(ws, doc, msg);
+
+			expect(sendSpy).toHaveBeenCalledTimes(0);
+
+			ws.close();
+			sendSpy.mockRestore();
+		});
+	});
+
+	describe('when trying to close already closed connection', () => {
+		const setup = async () => {
+			ws = await TestHelper.setupWs(wsUrl);
+
+			jest.spyOn(ws, 'close').mockImplementationOnce(() => {
+				throw new Error('some error');
+			});
+		};
+
+		it('should throw error when trying to close already closed connection', async () => {
+			await setup();
+			try {
+				const doc: { conns: Map<WebSocket, Set<number>> } = { conns: new Map() };
+				service.closeConn(doc as WsSharedDocDo, ws);
+			} catch (err) {
+				expect(err).toBeDefined();
+			}
+
+			ws.close();
+		});
+	});
+
+	describe('when ping failed', () => {
+		const setup = async () => {
+			ws = await TestHelper.setupWs(wsUrl, 'TEST');
+
+			const messageHandlerSpy = jest.spyOn(service, 'messageHandler').mockImplementationOnce(() => {});
+			const closeConnSpy = jest.spyOn(service, 'closeConn');
+			jest.spyOn(ws, 'ping').mockImplementationOnce(() => {
+				throw new Error('error');
+			});
+
+			return {
+				messageHandlerSpy,
+				closeConnSpy,
+			};
+		};
+
+		it('should close connection', async () => {
+			const { messageHandlerSpy, closeConnSpy } = await setup();
+
+			service.setupWSConnection(ws);
+
+			await delay(200);
+
+			expect(closeConnSpy).toHaveBeenCalled();
+
+			ws.close();
+			messageHandlerSpy.mockRestore();
+			closeConnSpy.mockRestore();
+		});
+	});
+
+	describe('when awareness states size greater then one', () => {
+		const setup = async () => {
+			ws = await TestHelper.setupWs(wsUrl, 'TEST');
+
+			const doc = new WsSharedDocDo('TEST', service);
+			doc.awareness.states = new Map();
+			doc.awareness.states.set(1, ['test1']);
+			doc.awareness.states.set(2, ['test2']);
+
+			const messageHandlerSpy = jest.spyOn(service, 'messageHandler').mockImplementationOnce(() => {});
+			const sendSpy = jest.spyOn(service, 'send');
+			const getYDocSpy = jest.spyOn(service, 'getYDoc').mockImplementationOnce(() => doc);
+			const { msg } = createMessage([0, 1]);
+			jest.spyOn(AwarenessProtocol, 'encodeAwarenessUpdate').mockImplementationOnce(() => msg);
+
+			return {
+				messageHandlerSpy,
+				sendSpy,
+				getYDocSpy,
+			};
+		};
+
+		it('should send if awareness states size greater then one', async () => {
+			const { messageHandlerSpy, sendSpy, getYDocSpy } = await setup();
+
+			service.setupWSConnection(ws);
+
+			await delay(200);
+
+			expect(sendSpy).toHaveBeenCalledTimes(3);
+
+			ws.close();
+			messageHandlerSpy.mockRestore();
+			sendSpy.mockRestore();
+			getYDocSpy.mockRestore();
+		});
+	});
```

Comment:
I personally think its fine to not test updateDocument and flushDocument here since there is nothing going on there. But there should at least be some kind of testcoverage via an integration test for them.

Reply:
As i responded in other comment, we have an issue to do such integration tests right now, but I've added tests for updateDocument and flushDocument